include/wx/translation.h

```diff
@@ -331,13 +331,13 @@ inline const wxString& wxGetTranslation(const char *str1,
 #if !defined(WXINTL_NO_GETTEXT_MACRO)
     #if !defined(_)
 #ifndef wxNO_IMPLICIT_WXSTRING_ENCODING
-        #define _(s)                 (s)
+        #define _(s)                 wxString(s)
 #else
         #define _(s)                 wxASCII_STR(s)
 #endif
     #endif
     #define wxPLURAL(sing, plur, n)  ((n) == 1 ? (sing) : (plur))
-    #define wxGETTEXT_IN_CONTEXT(c, s)                     (s)
+    #define wxGETTEXT_IN_CONTEXT(c, s)                     _(s)
```

Comment:
I'm not sure if this one is right though as it makes this macro inconsistent with `wxGETTEXT_IN_CONTEXT_PLURAL`.

Again, perhaps it would help to see what errors exactly the current version results in?

Reply:
[`menu_osx.cpp`](https://github.com/wxWidgets/wxWidgets/blob/master/src/osx/menu_osx.cpp) is using it as wxString, like `wxGETTEXT_IN_CONTEXT(c, s) + "foo"`.
So it should be a string. or need to fix `menu_osx.cpp`.

Reply:
I think it would be better (and very certainly safer) to just fix this there.

In fact, I think the code is wrong anyhow because it doesn't translate `Ctrl`, i.e. instead of
```
wxGETTEXT_IN_CONTEXT("macOS menu item", "Preferences...") + "\tCtrl+,"
```
it should really be just
```
_("Preferences...\tCtrtl+,")
```
which would remove the need for using context completely. @lanurmi am I missing something here?

But to avoid any extensive changes in this PR you could just change the above to
```
wxGETTEXT_IN_CONTEXT("macOS menu item", "Preferences...") + wxString("\tCtrl+,")
```
for now.

Reply:
> In fact, I think the code is wrong anyhow because it doesn't translate `Ctrl`, i.e. instead of
> 
> ```
> wxGETTEXT_IN_CONTEXT("macOS menu item", "Preferences...") + "\tCtrl+,"
> ```
> 
> it should really be just
> 
> ```
> _("Preferences...\tCtrtl+,")
> ```
> 
> which would remove the need for using context completely. @lanurmi am I missing something here?

First of all, I suspect this particular string is never actually shown to users because of #16888. But let's ignore that for a while...

I don't think you would (or at least I would) want someone to be able change the keyboard shortcut something else than `Ctrl+,` in the standard macOS application menu, because that would mean any wxOSX application running with such a translation in use, would have a completely unusual shortcut for entering preferences.

The context also serves the purpose of documenting that this string's translation should match Apple's terminology.

Also, would the shortcut even work if `Ctrl` is translated? I tested modifying in the `internat` sample's first menu item's shortcut from `Ctrl-T` to `Strg-T` in the translation file, and running the sample in German (Germany). The result is that the shortcut becomes just `T`, and indeed just pressing `T` without any modifiers triggers the menu item.

Reply:
I understand.
So, if you will adopt lanurmi's opinion, I'll replace
`wxPLURAL(sing, plur, n)  ((n) == 1 ? (sing) : (plur))`
with
`wxPLURAL(sing, plur, n)  ((n) == 1 ? _(sing) : _(plur))`.

If not, I'll remove the changes from `translation.h` and replace `wxGETTEXT_IN_CONTEXT(c, s) + "foo"`
with
`wxGETTEXT_IN_CONTEXT(c, s) + wxASCII_STR("foo")`.
(wxASCII_STR might be better to support the no implicit conversions option?)

I'll follow your decision.

Reply:
> IMHO it would be highly confusing to say `_(X)` when you really mean `wxString(X)`, even though in this very context they happen to be equal.

`wxPLURAL` and `_(X)` are the macros for translation.
So, it's not coincidence. (at least, for me)
`((n) == 1 ? _(sing) : _(plur))` will work even when `wxUSE_INTL` is on.

Well, I noticed wxASCII_STR might be enough for constructing the key strings.
So, this might be better?
```
    #if !defined(_)
        #define _(s)                 wxASCII_STR(s)
    #endif
    #define wxPLURAL(sing, plur, n)  wxASCII_STR((n) == 1 ? (sing) : (plur))
    #define wxGETTEXT_IN_CONTEXT(c, s)                     wxASCII_STR(s)
```

Reply:
> I don't think you would (or at least I would) want someone to be able change the keyboard shortcut something else than `Ctrl+,` in the standard macOS application menu, because that would mean any wxOSX application running with such a translation in use, would have a completely unusual shortcut for entering preferences.

The idea is that `Ctrl` should be translated, not changing the shortcut. So we could also use something like

```cpp
wxString::Format("%s\t%s+,",
    wxGETTEXT_IN_CONTEXT("macOS menu item", "Preferences..."),
    _("Ctrl"));
```

I guess.

> The context also serves the purpose of documenting that this string's translation should match Apple's terminology.

OK, this is a good point.

> Also, would the shortcut even work if `Ctrl` is translated?

Well, I don't know if it does right now, but it very clearly should because surely German users should see `Strg+,` in the actual menu, shouldn't they?

> I tested modifying in the `internat` sample's first menu item's shortcut from `Ctrl-T` to `Strg-T` in the translation file, and running the sample in German (Germany). The result is that the shortcut becomes just `T`, and indeed just pressing `T` without any modifiers triggers the menu item.

This is a bug :-( I'm pretty sure this does work correctly in the other ports.

Reply:
@matyalatte Sorry for the confusion and all the unrelated deviations, but to address your question: I think we should keep the header unchanged and use `Format("%s...")` as shown above instead of string concatenation.

Reply:
> > Also, would the shortcut even work if `Ctrl` is translated?
> 
> Well, I don't know if it does right now, but it very clearly should because surely German users should see `Strg+,` in the actual menu, shouldn't they?

In the case of wxOSX, `Ctrl` really means `Cmd` and they should be seeing `⌘,`. But aside from that, yes, I agree with you; they should be seeing `Strg+,`. But could it be normalized and localized at the level of `wxMenu`, even if it was given untranslated?


> > I tested modifying in the `internat` sample's first menu item's shortcut from `Ctrl-T` to `Strg-T` in the translation file, and running the sample in German (Germany). The result is that the shortcut becomes just `T`, and indeed just pressing `T` without any modifiers triggers the menu item.
> 
> This is a bug :-( I'm pretty sure this does work correctly in the other ports.

Not sure if this is because the `internat` sample does not have access to `wxstd` translations on macOS, but even if that is the reason, it seems a bit fragile that an unknown modifier name is interpreted as nothing, and no assert fails.